### PR TITLE
subdag modifier + tag_outputs + refactor

### PR DESCRIPTION
**Pull request by elijahbenizzy**
[Short description explaining the high-level reason for the pull request]

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
This PR has turned into a massive one due to a rebase. It's replacing:
1. https://github.com/stitchfix/hamilton/pull/199
2. https://github.com/stitchfix/hamilton/pull/198

And will likely replace:
3. https://github.com/stitchfix/hamilton/pull/227

File: decorators.md [hunk 1]
```diff
@@ -365,3 +365,68 @@ two validators, one that checks the datatype of the series, and one that checks
...
+@extract_columns("timestamp", "user_id", "region")  # one of "US", "CA" (canada)
+def website_interactions(random_seed: int) -> pd.DataFrame:
+    return ...
+
+def interactions_filtered(filtered_interactions: pd.DataFrame, region: str) -> pd.DataFrame:
+    pass
+
+def unique_users(filtered_interactions: pd.DataFrame, grain: str) -> pd.Series:
+    """Gives the number of shares traded by the frequency"""
+    return ...
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply decorators.md:395]: Sorry this piece is a WIP

**Comment by skrawcz** [comment]
Some thoughts we should sync up on:
 - What's the long term story here from a code management perspective? 
 - How will manage new decorators? 
 - What should people be importing?

**Comment by elijahbenizzy** [comment]
TODO prior to general release:
- [ ] Some local testing
- [ ] Release RC candidate with https://github.com/stitchfix/hamilton/pull/238 in it
- [ ] Add more documentation/fix up examples for this
- [ ] Get general community approval for these changes

File: hamilton/function_modifiers_base.py [hunk 2]
```diff
@@ -1,363 +1,4 @@
...
-    for dag_modifier in node_transformers:
-        nodes = dag_modifier.transform_dag(nodes, config, fn)
-    node_decorators = getattr(fn, NodeDecorator.get_lifecycle_name(), [DefaultNodeDecorator()])
-    for node_decorator in node_decorators:
-        nodes = node_decorator.transform_dag(nodes, config, fn)
-    return nodes
+# Quick hack to make this a backwards compatible refactor
+# This allows us to define everything within the function_modifiers directory, and just refer to that
+# While maintaining old imports
+from hamilton.function_modifiers.base import *  # noqa F403
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/function_modifiers_base.py:4]: probably want something like this still?

**Review by skrawcz — APPROVED** [review]
Looks good.

File: hamilton/function_modifiers_base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply hamilton/function_modifiers_base.py:4]: Added it but this is not a public facing API

**Pull request merged**